src/Npgsql/NpgsqlConnector.cs

```diff
@@ -607,32 +607,30 @@ async Task RawOpen(NpgsqlTimeout timeout, bool async, CancellationToken cancella
                         var certPath = Settings.ClientCertificate ?? PostgresEnvironment.SslCert;
-                        var certPathExists = true;
 
-                        if (certPath is null)
+                        if (certPath is null && PostgresEnvironment.SslCertDefault is { } certPathDefault && File.Exists(certPathDefault))
+                            certPath = certPathDefault;
+
+                        if (certPath != null)
                         {
-                            certPath = PostgresEnvironment.SslCertDefault;
-                            certPathExists = File.Exists(certPath);
-                        }
+                            var certKey = Settings.ClientCertificateKey ?? PostgresEnvironment.SslKey;
+                            var cert = new X509Certificate2(certPath, certKey);
 
-                        if (certPathExists)
-                            clientCertificates.Add(new X509Certificate(certPath!));
+                            clientCertificates.Add(cert);
+                        }
 
                         ProvideClientCertificatesCallback?.Invoke(clientCertificates);
 
-                        RemoteCertificateValidationCallback certificateValidationCallback;
-                        if (Settings.TrustServerCertificate)
-                            certificateValidationCallback = (sender, certificate, chain, errors) => true;
-                        else if (UserCertificateValidationCallback != null)
-                            certificateValidationCallback = UserCertificateValidationCallback;
-                        else
-                            certificateValidationCallback = DefaultUserCertificateValidationCallback;
-
+                        var certificateValidationCallback =
+                            (Settings.TrustServerCertificate) ? SslTrustServerValidation :
+                            (Settings.RootCertificate ?? PostgresEnvironment.SslCertRoot) is { } certRootPath ? SslRootValidation(certRootPath) :
+                            (UserCertificateValidationCallback is { } userValidation) ? userValidation : SslDefaultValidation;
                         var sslStream = new SslStream(_stream, leaveInnerStreamOpen: false, certificateValidationCallback);
-                        if (async)
-                            await sslStream.AuthenticateAsClientAsync(Host, clientCertificates, SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, Settings.CheckCertificateRevocation);
-                        else
-                            sslStream.AuthenticateAsClient(Host, clientCertificates, SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, Settings.CheckCertificateRevocation);
+
+                        await sslStream
+                            .AuthenticateAsClientAsync(Host, clientCertificates, SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, Settings.CheckCertificateRevocation)
+                            .ConfigureAwait(!async);
```

Comment:
Because we a going to drop the scope, so we could start doing changes.

Reply:
Please, let's do that separately in a global PR. It could be dangerous (and definitely unclear) if we have two different mechanisms for this within the same codebase. For example, ConfigureAwait(true) here looks like it preserves the synchronization context, but up the stack NoSynchronizationContextScope has already set it to null...

Reply:
Leave only the async path without `ConfigureAwait`?

Reply:
BTW I realized this is worse - the PR changes the code to always call AuthenticateAsClientAsync, if when we're in the sync path (this is sync-over-async)... We shouldn't do this regardless of what we do with ConfigureAwait afterwards - or am I misunderstand something?

I think we should just go back to how it was before - call either AuthenticateAsClientAsync or AuthenticateAsClient based on the async variable.

Reply:
Okay, will revert.

Reply:
Thanks!